### ISSUE TITLE
windows: run cannot find *.cmd

### ISSUE THREAD

**Issue by emillon** [ocaml/dune#9128]
When a program has an extension `.cmd` (ex: `yarn.cmd`), `(run)` does not find it. I think that's because we try to add `.exe` but instead we should consider all extensions in `PATHEXT`.

**Comment by nojb** [comment]
I assume `yarn.cmd` is a CMD script. Regardless of the extension, is it possible to execute it using `run` at all?

**Comment by emillon** [comment]
I don't know, but since typing `yarn` in a shell will run it, I think it's expected that `run` should execute it. Do you mean that we shouldn't have this expectation, or that in the current state of affaires `run` can not execute such programs?

**Comment by nojb** [comment]
> Do you mean that we shouldn't have this expectation, or that in the current state of affaires `run` can not execute such programs?

Well, personally, I don't have the expectation. In general script files are not executable programs by themselves (they are just text files) so they cannot be directly executed, they need to be passed to an interpreter program to run (eg `bash` or `cmd.exe`). On Linux the distinction is blurred because of the `#!` which allows running scripts as if they were executables. To be checked, but my recollection is that `CreateProcess` will **not** execute a script passed as argument.

**Comment by emillon** [comment]
OK, that's interesting. I thought that `PATHEXT` corresponded to programs that are compatible with `CreateProcess`. How does Windows (for example, explorer) know how to run a particular file extension? is it just the file extension association API through the registry?

Also there's still something to do for things like `.com` files though? I don't think that they are supported by `run` at the moment but `CreateProcess` should work on them.

**Comment by nojb** [comment]
> I thought that `PATHEXT` corresponded to programs that are compatible with `CreateProcess`

I don't know anything about `PATHEXT`, but the fact that it can be modified by the user makes me think that it cannot correspond to what is understood by `CreateProcess` (which is fixed in stone).

A quick internet search finds the explanation below. `PATHEXT` seems to be some mechanism at the level of the shell to execute scripts.

https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.3

![image](https://github.com/ocaml/dune/assets/113560/df3f841d-320a-467c-81e2-4c24e6d29247)

**Comment by rgrinberg** [comment]
> > Do you mean that we shouldn't have this expectation, or that in the current state of affaires `run` can not execute such programs?
> 
> Well, personally, I don't have the expectation. In general script files are not executable programs by themselves (they are just text files) so they cannot be directly executed, they need to be passed to an interpreter program to run (eg `bash` or `cmd.exe`). On Linux the distinction is blurred because of the `#!` which allows running scripts as if they were executables. To be checked, but my recollection is that `CreateProcess` will **not** execute a script passed as argument.

That might be true, but I find it to be quite annoying in practice and the cause of a non trivial amount of dune builds not being portable. I don't know see why `run` should necessarily correspond to what `CreateProcess` does on windows. If there's really a need for something that low level, we could always introduce something lower level than `run`.

**Comment by nojb** [comment]
> I don't know see why `run` should necessarily correspond to what `CreateProcess` does on windows.

Do you have an alternative suggestion? Going through the shell is going to create non-trivial difficulties with quoting, and also will impact efficiency rather negatively.

**Comment by nojb** [comment]
> I don't know see why `run` should necessarily correspond to what `CreateProcess` does on windows.

And by the way, isn't this the point of `(system ...)` after all? Can this be used to run programs and scripts both on Linux and Windows? (I haven't used it much myself, maybe it doesn't work seamlessly; in which case maybe we should focus our improvements there.)

**Comment by emillon** [comment]
Rudi raises a good point: you can have a `dune` file with `(run yarn)` which works perfectly well on Linux, but because on Windows `yarn` is implemented as a `.cmd` file, you need to switch it to `(system yarn)` on Linux as well. That brings a fairly low level concern to something as high level as a `dune` file.

> Do you have an alternative suggestion? Going through the shell is going to create non-trivial difficulties with quoting, and also will impact efficiency rather negatively.

I'm curious to see if there's a way to have `(run)` query how `.cmd` files are executed, and have it call `CreateProcess` with the right arguments. In a way, that would be like emulating `#!` lines.

**Comment by rgrinberg** [comment]
> Do you have an alternative suggestion? Going through the shell is going to create non-trivial difficulties with quoting, and also will impact efficiency rather negatively.

I think we could just consult this `PATHEXT` environment variable to determine what should be executed as a script

> And by the way, isn't this the point of (system ...) after all? Can this be used to run executables and scripts both on Linux and Windows? (I haven't used it much myself, maybe it doesn't work seamlessly; in which case maybe we should focus our improvements there.)

Using `system` should be acceptable as well. Although it seems unfortunate that everyone will have to remember which binaries are provided as commands on Windows to make dune files portable.

**Comment by nojb** [comment]
> I think we could just consult this `PATHEXT` environment variable to determine what should be executed as a script

I don't think it is that simple because you also need to know which interpreter to use for each extension. Maybe an ad-hoc special case in `(run)` for `.bat` and `.cmd` (which would be passed to `cmd.exe` instead of executed directly) could suffice.

**Comment by dra27** [comment]
PATHEXT tells the command processor (cmd) what it should consider as executable extensions, and it special-cases (I think) com, exe, bat and cmd. The rest are left to ShellExecuteEx, which uses the registry to figure out what to do. That’s how you make Perl scripts, etc. executable from the command prompt. It doesn’t expose you to cmd’s quoting, etc - it’s just that a different system call should be used IIRC
